bohr/snorkel_utils.py

```diff
@@ -1,20 +1,334 @@
+        return len(self.__issues)
+
+    def __getitem__(self, idx) -> Issue:
+        return self.__issues[idx]
+
+    def match_label(self, stemmed_labels: Set[str]) -> bool:
+        for issue in self.__issues:
+            if not issue.stemmed_labels.isdisjoint(stemmed_labels): return True
+        return False
+
+    def contains_label(self, stemmed_label: str) -> bool:
+        for issue in self.__issues:
+            if stemmed_label in issue.stemmed_labels: return True
+        return False
+
+    def match(self, stemmed_keywords: Set[str]) -> bool:
+        for issue in self.__issues:
+            if not issue.stems.isdisjoint(stemmed_keywords): return True
+        return False
+
+    def contains(self, stemmed_keyword: str) -> bool:
+        for issue in self.__issues:
+            if stemmed_keyword in issue.stems: return True
+        return False
+
+    def match_bigram(self, stemmed_bigrams: Set[Tuple[str, str]]) -> bool:
+        for issue in self.__issues:
+            if not issue.stem_bigrams.isdisjoint(stemmed_bigrams): return True
+        return False
+
+    def contains_bigram(self, stemmed_bigram: Tuple[str, str]) -> bool:
+        for issue in self.__issues:
+            if stemmed_bigram in issue.stem_bigrams: return True
+        return False
+
+
+
+@dataclass
+class CommitFile:
+    filename: str
+    status: str
+    patch: Optional[str]
+    changes: Optional[str]
+
+
+class CommitFiles:
+    def __init__(self, files):
+        self.__files = files
+
+    def __len__(self) -> int:
+        return len(self.__files)
+
+    def __getitem__(self, idx) -> CommitFile:
+        return self.__files[idx]
+
+@dataclass
+class CommitMessage:
+    raw: str
+
+    @cached_property
+    def tokens(self) -> Set[str]:
+        if self.raw is None: return set()
+        return safe_word_tokenize(self.raw)
+
+    @cached_property
+    def ordered_stems(self) -> List[str]:
+        stemmer = PorterStemmer()
+        return [stemmer.stem(w) for w in self.tokens]
+
+    @cached_property
+    def stems(self) -> Set[str]:
+        return set(self.ordered_stems)
+
+    @cached_property
+    def stem_bigrams(self) -> Set[Tuple[str, str]]:
+        return set(bigrams(self.ordered_stems))
+
+    def match(self, stemmed_keywords: Set[str]) -> bool:
+        return not self.stems.isdisjoint(stemmed_keywords)
+
+    def match_bigram(self, stemmed_bigrams: Set[Tuple[str, str]]) -> bool:
+        return not self.stem_bigrams.isdisjoint(stemmed_bigrams)
+
+
+
+@dataclass
+class Commit:
+
+    owner: str
+    repository: str
+    sha: str
+    raw_message: str
+    message: CommitMessage = field(init=False)
+
+    def __post_init__(self):
+        self.message = CommitMessage(self.raw_message)
+
+    @lru_cache(maxsize=4)
+    def __load_df(self, type: str, owner: str, repository: str):
+        path = TRAIN_DIR / type  / owner / f"{repository}.csv"
+        if path.is_file():
+            return pd.read_csv(path, index_col=['sha'])
```

Comment:
If there are no labels, pandas reads the empty CSV cell as `float` `(NaN)`, which later causes problems when splitting by `,`